File: leetcode/134.gas-station/go/gasstation.go
package gasstation

func canCompleteCircuit(gas []int, cost []int) int {
	if len(gas) == 0 {
		return 0
	}

	total := 0
	res := 0
	curr := 0

	for i := range gas {
		v := gas[i] - cost[i]
		total += v
		curr += v
		if curr < 0 {
			res = i + 1
			curr = 0
		}
	}

	if total < 0 {
		return -1
	}

	return res
}

File: leetcode/134.gas-station/cpp/gasstation.cpp
#include <vector>

class Solution
{
public:
    int canCompleteCircuit(std::vector<int> &gas, std::vector<int> &cost)
    {
        int res = 0;
        int total = 0;
        int curr = 0;

        for (int i = 0; i < gas.size(); i++)
        {
            int v = gas[i] - cost[i];
            total += v;
            curr += v;

            if (curr < 0)
            {
                curr = 0;
                res = i + 1;
            }
        }

        if (total < 0)
            return -1;
        return res;
    }
};
